### PR TITLE
LPS-119818  Compare options by its value

### PR DESCRIPTION
**Pull request by alinedoleron**
https://issues.liferay.com/browse/LPS-119818

### CODE REVIEW

**Comment by alinedoleron** [comment]
ci:test:sf

**Comment by alinedoleron** [comment]
ci:test:relevant

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 1]
```diff
@@ -140,6 +206,25 @@ const syncActions = (pages, actions) => {
 				.filter((key) => !targetFieldExists(outputs[key], pages))
 				.map((key) => delete outputs[key]);
 		}
+		else if (action.action === 'calculate') {
+			const expressionFields = getExpressionFields(action);
+
+			if (expressionFields && expressionFields.length > 0) {
+				expressionFields.forEach((field) => {
+					if (!targetFieldExists(field, pages)) {
+						const inexistentField = new RegExp(field, 'g');
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:271]: @alinedoleron, please add a break line after this line. Ideally, we should separate variable declarations from statements. 😉

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/Sidebar/Sidebar.es.js [hunk 2]
```diff
@@ -728,7 +729,6 @@ class Sidebar extends Component {
 			});
 		}
 	}
-
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/Sidebar/Sidebar.es.js:731]: @alinedoleron, why did you remove this line?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/css/main.scss [hunk 3]
```diff
@@ -277,6 +277,13 @@
 	.modal-dialog {
 		position: relative;
 	}
+
+	.lfr-ddm-forms-delete-rule {
+		.modal-item-last button.btn-primary {
+			background-color: #da1414;
+			border-color: #da1414;
+		}
+	}
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/css/main.scss:286]: @alinedoleron, add `.lfr-ddm-forms-delete-rule` following alphabetic order.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js [hunk 4]
```diff
@@ -322,10 +324,48 @@ class LayoutProvider extends Component {
...
+					)}
+				/>
 			</span>
 		);
 	}
 
+	_handleDeleteFieldModalButtonClicked(event) {
+		const {activePage, fieldName} = this.refs.existingRuleModal.data;
+
+		if (event.target.classList.contains('btn-primary')) {
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js:364]: @alinedoleron, would be better check for the button's label? Check only for the class `btn-primary` isn't so intuitive for the code reader and also will require a rework from our side if this class changes.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js [hunk 5]
```diff
@@ -44,6 +45,12 @@ class RuleList extends Component {
 				true
 			)
 		);
+
+		var tooltip = ClayTooltip.init();
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js:51]: @alinedoleron, why are you using `var` instead of `let` or `const`?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js [hunk 6]
```diff
@@ -44,6 +45,12 @@ class RuleList extends Component {
...
+
+		setTimeout(() => {
+			tooltip.selectors = ['.invalid-rule'];
+		}, 1500);
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js:55]: @alinedoleron, can you please clarify why this is necessary?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.soy [hunk 7]
```diff
@@ -24,7 +24,8 @@
 				outputLabel: ?
 			]>,
 			logicalOperator: string,
-			rulesCardOptions: ?
+			rulesCardOptions: ?,
+			invalidRule: ?
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.soy:28]: @alinedoleron, can you order it alphabetically?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.soy [hunk 8]
```diff
@@ -71,7 +72,8 @@
...
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.soy:76]: @alinedoleron, can you order it alphabetically?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.soy [hunk 9]
```diff
@@ -142,7 +144,23 @@
 						</div>
 
 						<div class="list-group-item-field">
+							{let $titleMessage kind="text"}
+								{msg desc=""}due-to-missing-fields{/msg}
+							{/let}
+							{let $labelMessage kind="text"}
+								{msg desc=""}broken-rule{/msg}
+							{/let}
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.soy:152]: @alinedoleron, these variables should be declared inside the `if`, once they'll only be used if the rule is invalid.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-lang/src/main/resources/content/Language.properties [hunk 10]
```diff
@@ -1,3 +1,4 @@
+a-rule-is-applied-to-this-field=A rule is applied to this field. By deleting it, it will also break the rule that is using it.
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-lang/src/main/resources/content/Language.properties:1]: @alinedoleron, please separate this commit on 2 commits, one just adding the missing language keys and another one with the result of the `buildLang` command.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 11]
```diff
@@ -59,6 +59,19 @@ const clearAllConditionFieldValues = (condition) => {
...
+
+	const field = visitor.findField((field) => {
+		return field.fieldName === fieldName;
+	});
+
+	options = field ? field.options : [];
+
+	return options;
+}
+
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js]: @alinedoleron, why was it necessary?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 12]
```diff
@@ -98,6 +87,22 @@ const formatRules = (pages, rules) => {
...
+				pages
+			);
+
+			if (
+				firstOperandFieldExists &&
+				fieldWithOptions(firstOperandFieldType) &&
+				condition.operands[1].type != 'field'
+			) {
+				const fieldName = condition.operands[0].value;
+				const options = getFieldOptions(fieldName, pages);
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:101]: @alinedoleron, please add a break line after this line. Ideally, we should separate variable declarations from statements. 😉

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 13]
```diff
@@ -206,6 +226,29 @@ const getExpressionFields = (action, regex = regexFieldPattern) => {
...
+	});
+
+	options = field ? field.options : [];
+
+	return options;
+};
+
+const getFieldType = (fieldName, pages) => {
+	return getFieldProperty(pages, fieldName, 'type');
+};
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:244]: @alinedoleron, a very [similar method](https://github.com/natocesarrego/liferay-portal/blob/65c170e3780026a0d25b48b4f8a697af5bccade4/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleEditor/RuleEditor.es.js#L598-L602) is available on `RuleEditor`, should we reuse it?

Please, analyze if makes sense reuse it here, as you did with the method `getFieldOptions`. Think if makes more sense the method be on `RuleEditor` or on `RuleSupport`. What is the purpose of these the classes? Consider it on your decision.

**Review by natocesarrego — CHANGES_REQUESTED** [review]
@alinedoleron, I needed to do many comments, the majority of them were related to source format, please take a look to learn from this pull. The remaining comments are related to the solution, please check them too. If you prefer, we can discuss more about this pull on a call.

Thanks for your hard work on this pull. 🙂

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js (hunk 5 above)
Comment (alinedoleron) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js:51]: I didn't noticed. It was a copy and paste from a piece of code of Clay's demo.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js (hunk 6 above)
Comment (alinedoleron) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleList/RuleList.es.js:55]: I thought it was necessary for the tooltip appears smoothly, like it is done on [Clay's demo](https://github.com/liferay/clay/blob/2.x/packages/clay-tooltip/demos/index.html). But I removed to test now and it behaves the same way as before. So it's not necessary indeed. Thank you.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js (hunk 11 above)
Comment (alinedoleron) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js]: This method is used both on RulesSupport (inside formatRules method) and RuleEditor (inside prepareStateForRender method).

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js (hunk 13 above)
Comment (alinedoleron) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:244]: I think it's better to keep it in RuleSupport, since it's also being used on RuleEditor and RuleList.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js (hunk 4 above)
Comment (alinedoleron) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js:364]: I have a question. Should the button labels be localisable? I thought when we changed the locale on `Control-panel > Instance settings > Localization` the label should be according to the language. That was the reason I didn't compare with the label.

**Comment by alinedoleron** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js (hunk 4 above)
Comment (natocesarrego) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js:364]: Yes. Now, I see your point, @alinedoleron. Does this button has a unique identifier (`id`) which we can use? Using this strategy would make this code stronger against CSS/UI changes. 🙂

Comment (alinedoleron) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/LayoutProvider/LayoutProvider.es.js:364]: @natocesarrego, actually, after talking to Matu right now, he said the first solution I has done is the right. He said also there's no way to put a unique identifier on the footersButtons. I asked him about the class changes, but he said it's very hard this to be changed, as it's a pattern on the web. As an alternative, he said I could try to verify if the modal had close after the click, but it didn't work. For both cases after clicking on the buttons the modal keeps opened before the verification. So, I'll change it back to the code before.
